apps/mobile/src/store/key-store.ts

```diff
@@ -109,6 +127,59 @@ export function useKeyStore() {
       const nextAccountIndex =
         fingerprintAccounts.length === 0 ? 0 : highestKeychainAccountIndex + 1;
 
+      // in extension secretKey is the mnemonic
+      const secretKey = mnemonic;
+
+      const legacyAccountActivityLookupGaia =
+        await checkForLegacyGaiaConfigWithKnownGeneratedAccountIndex(secretKey);
+
+      // hardcoding this to match extension count for dev wallet
+      // checkForLegacyGaiaConfigWithKnownGeneratedAccountIndex fails with
+      // [TypeError: _this.nodePbkdf2 is not a function (it is undefined)]
+      const legacyAccountActivityLookup =
+        legacyAccountActivityLookupGaia === 0 ? 123 : legacyAccountActivityLookupGaia;
+
+      async function doesStacksAddressHaveBalance(address: string) {
+        const controller = new AbortController();
+        const resp = await stxClient.getAccountBalance(address, controller.signal);
+        return Number(resp.stx.balance) > 0;
+      }
+
+      try {
+        // I think index here is not incrementing properly and it's needed by getStacksAddressByIndex
+        void recurseAccountsForActivity({
+          async doesAddressHaveActivityFn(index: number) {
+            const stxAddress = getStacksAddressByIndex(
+              secretKey,
+              AddressVersion.MainnetSingleSig
+            )(index);
+            // here we call doesStacksAddressHaveBalance which calls stacks client directly not using react query
+            const hasStxBalance = await doesStacksAddressHaveBalance(stxAddress);
+            // const hasNames = await doesStacksAddressHaveBnsName(stxAddress);
+            // const btcAddress = getNativeSegwitMainnetAddressFromMnemonic(secretKey)(index);
+            // const hasBtcBalance = await doesBitcoinAddressHaveBalance(btcAddress.address!);
+            // TODO: add inscription check here also?
+            return hasStxBalance; // || hasNames || hasBtcBalance;
+          },
+        }).then(recursiveActivityIndex => {
+          if (recursiveActivityIndex <= legacyAccountActivityLookup) return;
+          console.info('Found account activity at higher index', { recursiveActivityIndex });
+          // dispatch(stxChainSlice.actions.restoreAccountIndex(recursiveActivityIndex));
+          console.log('recursiveActivityIndex', recursiveActivityIndex);
+          this.createNewAccountOfWallet(fingerprint, recursiveActivityIndex);
```

Comment:
In the extension `stx-chain.slice.ts` for `stxChainSlice.actions.restoreAccountIndex(recursiveActivityIndex)` it is using `highestAccountIndex` which gets incremented in `createNewAccount` so maybe that's where I am going wrong.

Reply:
There should not be side effects in this callback.

The idea of this recursive look up is that we give it a callback to determine whether there's activity or not. Then, the response (which isn't used here because of `void recurseAccountsForActivity`) tell you the highest account index that has activity. 

Later, you use that index to know how many accounts should be created.

Reply:
I se here you're try to recurse as well, but that isn't needed because all the recursion is handled internally to this function, it's way easier that way

Reply:
I based this code on the extension code commented out above it. That was using `dispatch(stxChainSlice.actions.restoreAccountIndex(recursiveActivityIndex))` to restore accounts. 

I could be mistaken, but when I debug this, it seems like in `.then` the `recursiveActivityIndex` is always `0` so this code is never reached (on extension or copied version). See extension [draft](https://github.com/leather-io/extension/pull/5939/files) . 

I think accounts are actually created further on by:
```
 if (legacyAccountActivityLookup !== 0)
        // dispatch(stxChainSlice.actions.restoreAccountIndex(legacyAccountActivityLookup)); -> extension
        this.createNewAccountOfWallet(fingerprint, legacyAccountActivityLookup);
```

`legacyAccountActivityLookup` is determined by `checkForLegacyGaiaConfigWithKnownGeneratedAccountIndex` which I think is actually what is determining how many accounts to create. 

What I think I can get to work is to run `this.createNewAccountOfWallet` inside of the recursive function if there is a balance

Reply:
Why does it always return 0? Does the account have activity you're testing with? It's quite slow, so are you waiting until the end? 

<img width="761" alt="image" src="https://github.com/user-attachments/assets/2de85139-dac7-44ce-8722-e18e2e51e049">

> What I think I can get to work is to run this.createNewAccountOfWallet inside of the recursive function if there is a balance

I really don't think this is the way to do it. The purpose of `recurseAccountsForActivity` is to handle the recursion of the look up. If you have to do additional recursion outside that fn, then either your implementation misses the point of the function, or the implementation of `recurseAccountsForActivity` doesn't work and needs fixing.

Happy to run through in more detail the gist of this function, but really its just a look up function that stops after _n_ false values and returns a number.

Also, let's remove all the gaia stuff so we don't need to worry about legacy account look up code.

Reply:
Thanks Kyran. I will do more testing and then implement the best solution using your recommendations. 

I'm not sure if it's working as we think in the extension right now. I am testing with my personal wallet and the leather dev wallet and in both cases accounts are created only if `if (legacyAccountActivityLookup !== 0)`.

We have two calls to add accounts:
- in `recurseAccountsForActivity.then` 
`dispatch(stxChainSlice.actions.restoreAccountIndex(recursiveActivityIndex));`
- after a check for `if (legacyAccountActivityLookup !== 0)`:
`dispatch(stxChainSlice.actions.restoreAccountIndex(legacyAccountActivityLookup));`


- If I remove the `dispatch` after the `legacyAccountActivityLookup` check, no accounts are created apart from the first one. 
- If I add a debugger in `recurseAccountsForActivity.then`  it never gets hit. 

Here's the full code. 
```
 try {
      void recurseAccountsForActivity({
        async doesAddressHaveActivityFn(index) {
          ....
          return hasStxBalance || hasNames || hasBtcBalance;
        },
      }).then(recursiveActivityIndex => {
        if (recursiveActivityIndex <= legacyAccountActivityLookup) return;
        logger.info('Found account activity at higher index', { recursiveActivityIndex });
        dispatch(stxChainSlice.actions.restoreAccountIndex(recursiveActivityIndex));
      });
    } catch (e) {
      // Errors during account restore are non-critical and can fail silently
    }

 ....
    if (legacyAccountActivityLookup !== 0)
      dispatch(stxChainSlice.actions.restoreAccountIndex(legacyAccountActivityLookup));
  };
```

If I change this code to be:
```
const accountsWithBalances = recurseAccountsForActivity({
        async doesAddressHaveActivityFn(index) {
          ....
          return hasStxBalance || hasNames || hasBtcBalance;
        },
      })
```

`accountsWithBalances` is always `0`. If I add logs inside of there I can see it finding accounts with balances but I think it's not working as we expect.

Reply:
Upon further investigation into this, it seems like the recursion we are using does not generate any accounts. They are in fact all generated initially based on the count returned from `checkForLegacyGaiaConfigWithKnownGeneratedAccountIndex` 

Here is a demo PR from the extension showing that: https://github.com/leather-io/extension/pull/5945